### PR TITLE
Integration with visualization

### PR DESCRIPTION
**Pull request by kolayne**
Added integration with visualization

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,6 +22,13 @@ cmake .. -DCOMPILE_FOR_CPU=ON
 cmake --build . -- -j "$(nproc)"
 ```
 
+### Executing
+To run the application, you have to:
+
+1. Start a visualization app (we recommend [this one](https://github.com/physarumAdv/Physarum_webGL))
+
```
Comment (kolayne) [inline README.md:29]: Codacy found an issue: [[list-item-spacing] Extraneous new line after list item](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

**Comment by kolayne** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038&bid=20268787)

File: README.md [hunk 2]
```diff
@@ -22,6 +22,13 @@ cmake .. -DCOMPILE_FOR_CPU=ON
 cmake --build . -- -j "$(nproc)"
 ```
 
+### Executing
+To run the application, you have to:
+
+1. Start a visualization app (we recommend [this one](https://github.com/physarumAdv/Physarum_webGL))
```
Comment (kolayne) [inline README.md:28]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

File: README.md [hunk 3]
```diff
@@ -22,6 +22,13 @@ cmake .. -DCOMPILE_FOR_CPU=ON
...
+
+2. Create a directory `local` in the process's **working directory** with a file `visualization_endpoint.txt` inside it, containing a url which accepts the simulated data (an example is in [local/visualization_endpoint_example.txt](local/visualization_endpoint_example.txt))
```
Comment (kolayne) [inline README.md:30]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

**Comment by kolayne** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038&bid=20268787)

File: README.md (hunk 1 above)
Comment (kolayne) [inline README.md:29]: Codacy found an issue: [[list-item-spacing] Extraneous new line after list item](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

File: README.md [hunk 4]
```diff
@@ -22,6 +22,13 @@ cmake .. -DCOMPILE_FOR_CPU=ON
 cmake --build . -- -j "$(nproc)"
 ```
 
+### Executing
+To run the application, you have to:
+
+1. Start a visualization app (we recommend [this one](https://github.com/physarumAdv/Physarum_webGL))
+
+2. Create a directory `local` in the process's **working directory** with a file `visualization_endpoint.txt` inside it, containing a url which accepts the simulated data (an example is in [local/visualization_endpoint_example.txt](config/visualization_endpoint_example.txt))
```
Comment (kolayne) [inline README.md:30]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

File: README.md (hunk 2 above)
Comment (kolayne) [inline README.md:28]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

**Comment by kolayne** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038&bid=20268787)

File: README.md [hunk 5]
```diff
@@ -15,13 +15,22 @@ cmake ..
...
 ```bash
 mkdir cmake-build-debug && cd cmake-build-debug
 cmake .. -DCOMPILE_FOR_CPU=ON
 cmake --build . -- -j "$(nproc)"
 ```
 
+### Executing
+To run the application, you have to:
+
+1.  Start a visualization app (we recommend [this one](https://github.com/physarumAdv/Physarum_webGL))
```
Comment (kolayne) [inline README.md:29]: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6217038)

File: src/visualization_integration.cu [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    for(int i = 0; i < n_of_nodes; ++i)
+    {
+        if(!nodes[i].does_contain_particle())
+            continue;
+
+        SpacePoint coords = nodes[i].get_coordinates();
+        x.push_back(coords.x);
+        y.push_back(coords.y);
+        z.push_back(coords.z);
```
Comment (tanya-kta) [inline src/visualization_integration.cu]: If I understand this correctly, you send coordinates of nodes. But that should be coordinates of particle, shouldn't that?

**Review by tanya-kta — CHANGES_REQUESTED** [review]

File: src/visualization_integration.cu (hunk 6 above)
Comment (kolayne) [reply src/visualization_integration.cu]: Yes, you understand the meaning of the vectors correctly. We usually expect the coordinates of a node to not differ a lot from the coordinates of a particle located in it, but I think you are totally correct and it surely won't be worse if we use coordinates of _particles_ to tell where _particles_ are located, especially assuming that the distance between the neighbor node may vary

**Review by tanya-kta — APPROVED** [review]
That's good

**Pull request merged**